File: 2210-count-hills-and-valleys-in-an-array/2210-count-hills-and-valleys-in-an-array.py
class Solution:
    def countHillValley(self, nums: List[int]) -> int:
        # 步驟一：先消除連續的重複數值
        filtered = [nums[0]]
        for i in range(1, len(nums)):
            if nums[i] != nums[i - 1]:
                filtered.append(nums[i])

        # 步驟二：遍歷中間元素，檢查是否為 hill 或 valley
        count = 0
        for i in range(1, len(filtered) - 1):
            if filtered[i] > filtered[i - 1] and filtered[i] > filtered[i + 1]:
                count += 1 # Hill
            elif filtered[i] < filtered[i - 1] and filtered[i] < filtered[i + 1]:
                count += 1 # Valley
        
        return count

File: 2210-count-hills-and-valleys-in-an-array/2210-count-hills-and-valleys-in-an-array.cpp
class Solution {
public:
    int countHillValley(vector<int>& nums) {
        // 步驟一：先消除連續的重複數值
        vector<int> filtered;
        filtered.push_back(nums[0]);

        for(int i = 1; i < nums.size(); ++i){
            if (nums[i] != nums[i - 1]){
                filtered.push_back(nums[i]);
            }
        }
        // Step 2: Count hills and valleys
        int count = 0;
        for (int i = 1; i < filtered.size() - 1; ++i){
            if (filtered[i] > filtered[i - 1] && filtered[i] > filtered[i + 1]){
                count++;
            } else if (filtered[i] < filtered[i - 1] && filtered[i] < filtered[i + 1]){
                count++;
            }
        }
        return count;
    }
};